[Processor] Stream responses for go and python runtimes

rokatyy opened the pull request:
Consist of 3 PRs:

*  https://github.com/nuclio/nuclio/pull/3665
* https://github.com/nuclio/nuclio/pull/3671
* https://github.com/nuclio/nuclio/pull/3665
* https://github.com/nuclio/nuclio/pull/3674
* https://github.com/nuclio/nuclio/pull/3676

+ two commits fixing tests  https://github.com/nuclio/nuclio/pull/3678/commits/9140aeae231417dd4e7ac6a36e9f5ab7519f5158 https://github.com/nuclio/nuclio/pull/3678/commits/3e7c42713f8b0dc66fdd560e537d7e07962adeff

rokatyy committed:
[Processor] Support yielding in wrapper (#3670)

Jira - https://iguazio.atlassian.net/browse/NUC-465

Refactored handling of function outputs to support both batched and
streaming responses in a unified async flow.

Introduced `_generate_processor_packets()` to abstract packet generation
logic for different response types (single, batched, and streamed).

Replaced manual duration encoding and response writing with
_handle_entrypoint_output() that consumes the generated packets and
writes them to the processor socket.

Enables proper handling of async generators and streamed responses with
correct packet prefixes (`r`, `c`, `b`, `e`,`m`), improving flexibility
and consistency.

rokatyy committed:
[Dependencies] Bump nuclio-sdk-py to 0.6.0 (#3671)

Since nuclio-sdk changes were changed to be worth bumping minor, bumping
it here as well

rokatyy committed:
[Processor] Implement processing of stream (#3665)

Within this pr:

* added support of new types of messages from a wrapper. Namely:
  - 'c' for stream start
  - 'b' for streaming body
  - 'e' for end of stream
* resultChan will now contain object of result.Result interface which
may wrap nuclio.ProcessingResult under the hood. Those which have
`nuclio.ProcessingResult`also implement
`ResultWithNuclioProcessingResult`. The idea is that from worker we
always get `ResultWithNuclioProcessingResult` and from trigger we always
get `nuclio.ProcessingResult`. While all the logic that is before that
uses `result.Result`.
* also, moved TestEvent from test to a separate folder and not `test_*`
file, so it can be reused across the code, where we don't really care
about what is inside an event

Jira - https://iguazio.atlassian.net/browse/NUC-466

TomerShor commented:
Let's wait for https://github.com/nuclio/nuclio/pull/3676 and https://github.com/nuclio/nuclio/pull/3674 and merge them too at once

rokatyy committed:
[Processor] Support chunk timeout and always run watcher (#3674)

The timeout watcher now always runs, since stream processing occurs in a
separate goroutine. This ensures that if any error or timeout happens,
the connection is restarted to prevent residual data from previous
processing from leaking into the channel.

Additionally, improved connection status handling in the socket
allocator. Previously, we relied on the manager’s status, but since the
worker now may be released before the connection release finished (where
we set socket allocator manager status), we can't depend on that
anymore. Instead, the connection's own status is now checked to
determine readiness. This change applies only to the socket allocator,
which is the only one using non-blocking allocation.

A new method, SetStatusCodeFromError, was introduced to allow setting an
appropriate error code dynamically during stream processing—providing
users with clearer feedback when something goes wrong mid-stream.

Also added a new configuration parameter, `streamChunkTimeout`, which
behaves similarly to eventTimeout. The timeout watcher will now always
use the `smallest of streamChunkTimeout, eventTimeout`, or fall back to
a default of `10 seconds`.

Jira - https://iguazio.atlassian.net/browse/NUC-487

rokatyy committed:
[Processor] Add metric EventsStreamingFinishedSuccessfully (#3676)

Adds new metric `EventsStreamingFinishedSuccessfully` to signal about
successful processing of stream. If something happens during processing,
worker entity isn't aware of this, so it will be possible to find out by
checking that
`EventsStreamingStartedSuccessfully==EventsStreamingFinishedSuccessfully`.

Jira - https://iguazio.atlassian.net/browse/NUC-467

rokatyy commented:
@TomerShor merged

rokatyy committed:
Fix tests

rokatyy committed:
one more test fix

rokatyy commented:
`https://imagemagick.org/` website is down, so link checker fails, I assume it should be fixed soon as it was healthy just an hour ago

rokatyy commented:
benchmark report: https://github.com/nuclio/nuclio/actions/runs/15996491180
